File: Nemial.BrainGames/Games/Gcd.cs
using Nemial.BrainGames.Core;

namespace Nemial.BrainGames.Games;

public class Gcd : IGame
{
    private const string GameName = "Brain GCD";
    private const string GameDescription = "Find the greatest common divisor of given numbers.";

    private readonly GameEngine _engine = new();
    private readonly Random _random = new();

    public void Run()
    {
        var firstNum = _random.Next(1, 512);
        var secondNum = _random.Next(1, firstNum);
        var gcd = FindGcd(firstNum, secondNum);
        var question = $"{firstNum} {secondNum}";

        _engine.StartGame(GameName, GameDescription, question, gcd.ToString());
    }

    private static int FindGcd(int firstNum, int secondNum)
    {
        return firstNum % secondNum != 0 ? FindGcd(secondNum, firstNum % secondNum) : secondNum;
    }
}

File: src/games/gcd.rs
use crate::games::start_game;
use rand::Rng;

const GAME_NAME: &str = "Brain GCD";
const GAME_DESCRIPTION: &str = "Find the greatest common divisor of given numbers.";
pub fn start() {
    let mut rng = rand::thread_rng();

    let first_num: usize = rng.gen_range(1..256);
    let second_num: usize = rng.gen_range(1..first_num);

    let question = format!("{} {}", first_num, second_num);
    let answer = find_gcd(first_num, second_num);

    start_game(GAME_NAME, GAME_DESCRIPTION, answer.to_string(), question);
}

fn find_gcd(first_num: usize, second_num: usize) -> usize {
    if first_num % second_num != 0 {
        find_gcd(second_num, first_num % second_num)
    } else {
        second_num
    }
}
